lib/logging/index.js

```diff
@@ -0,0 +1,104 @@
+}
+
+function _getExporter() {
+  let {
+    kind,
+    logging: { exporter: loggingExporter },
+    credentials
+  } = cds.env.requires.telemetry
+
+  // for kind telemetry-to-otlp based on env vars
+  if (loggingExporter === 'env') {
+    const otlp_env = getEnvWithoutDefaults()
+    const dflt_env = getEnv()
+    const protocol =
+      otlp_env.OTEL_EXPORTER_OTLP_LOGS_PROTOCOL ??
+      otlp_env.OTEL_EXPORTER_OTLP_PROTOCOL ??
+      dflt_env.OTEL_EXPORTER_OTLP_LOGS_PROTOCOL ??
+      dflt_env.OTEL_EXPORTER_OTLP_PROTOCOL
+    loggingExporter = { module: _protocol2module[protocol], class: 'OTLPLogExporter' }
+  }
+
+  // use _require for better error message
+  const loggingExporterModule = _require(loggingExporter.module)
+  if (!loggingExporterModule[loggingExporter.class])
+    throw new Error(`Unknown logs exporter "${loggingExporter.class}" in module "${loggingExporter.module}"`)
+  const loggingConfig = { ...(loggingExporter.config || {}) }
+
+  if (kind.match(/to-cloud-logging$/)) {
+    if (!credentials) credentials = getCredsForCLSAsUPS()
+    if (!credentials) throw new Error('No SAP Cloud Logging credentials found.')
+    augmentCLCreds(credentials)
+    loggingConfig.url ??= credentials.url
+    loggingConfig.credentials ??= credentials.credentials
+  }
+
+  const exporter = new loggingExporterModule[loggingExporter.class](loggingConfig)
+  LOG._debug && LOG.debug('Using logs exporter:', exporter)
+
+  return exporter
+}
+
+module.exports = resource => {
+  if (!cds.env.requires.telemetry.logging?.exporter) return
+
+  const { logs, SeverityNumber } = require('@opentelemetry/api-logs')
+  const { LoggerProvider, BatchLogRecordProcessor } = require('@opentelemetry/sdk-logs')
+
+  let loggerProvider = logs.getLoggerProvider()
+  if (!loggerProvider.getDelegateLogger()) {
+    loggerProvider = new LoggerProvider({ resource })
+    logs.setGlobalLoggerProvider(loggerProvider)
+  } else {
+    LOG._warn && LOG.warn('LoggerProvider already initialized by a different module. It will be used as is.')
+    loggerProvider = loggerProvider.getDelegateLogger()
+  }
+
+  const exporter = _getExporter()
+  loggerProvider.addLogRecordProcessor(new BatchLogRecordProcessor(exporter))
+
+  const loggers = {}
+  const l2s = { 1: 'ERROR', 2: 'WARN', 3: 'INFO', 4: 'DEBUG', 5: 'TRACE' }
+
+  // intercept logs via format
+  const { format } = cds.log
+  cds.log.format = (module, level, ...args) => {
+    const res = format(module, level, ...args)
+
+    let log
+    try {
+      log = res.length === 1 && res[0].startsWith?.('{"') && JSON.parse(res[0])
+    } catch {
+      // ignore
+    }
+    if (log) {
+      const logger = loggers[module] || (loggers[module] = loggerProvider.getLogger(module))
+      const severity = l2s[level]
+      // TODO: what to log?
+      logger.emit({
+        severityNumber: SeverityNumber[severity],
+        severityText: severity,
+        body: log.msg,
+        attributes: { 'log.type': 'LogRecord' }
+      })
```

Comment:
to do

Reply:
check https://opentelemetry.io/docs/specs/otel/logs/data-model/#log-and-event-record-definition